# Phase 3: Data Loading, Flattening & Initial Inspection

**Objective:** Load the raw JSON data fetched from the Random User API, flatten the deeply nested structure into a Pandas DataFrame, and perform an initial health check to understand data types, missing values, and basic statistics.

**Business Context:** Understanding the raw structure helps us decide what needs cleaning before we can build our "User Profile Quality Score".


In [1]:
# Set the working directory to the project root
import os
import sys

# Get the absolute path of the current notebook's directory
notebook_dir = os.getcwd()
print(f"Current working directory: {notebook_dir}")

# Move up one level to the project root (since the notebook is inside 'notebooks/')
project_root = os.path.dirname(notebook_dir)
print(f"Project root directory: {project_root}")

# Change the working directory to the project root
os.chdir(project_root)

# Verify the change
print(f"New working directory: {os.getcwd()}")

if project_root not in sys.path:
    sys.path.insert(0, project_root)
    print("Project root added to Python path!")

Current working directory: C:\Users\Usman Khan\global-user-onboarding-analytics\notebooks
Project root directory: C:\Users\Usman Khan\global-user-onboarding-analytics
New working directory: C:\Users\Usman Khan\global-user-onboarding-analytics
Project root added to Python path!


In [3]:
# Importing the libraries 
import numpy as np
import pandas as pd
import json
from datetime import datetime

print(f"Pandas version: {pd.__version__}")

Pandas version: 3.0.5


In [7]:
# Defining the path to our raw data folder
raw_data_folder = "data/raw/"

#  Geting a list of all files in the raw data folder
all_files = os.listdir(raw_data_folder)

# Filter to only include JSON files
json_files = [] # Step 1: Creating an empty list

for f in all_files: # Step 2: Look at every file name in all_files one by one
    if f.endswith(".json"): # Step 3: Check if the file name ends with '.json'
        json_files.append(f) # Step 4: If yes, add it to our new list
        
# Sorting them so the newest one is last, and take the last one
if json_files:
    latest_json_file = sorted(json_files)[-1]
else:
    print("No JSON files found!")

# Build the full file path
file_path = os.path.join(raw_data_folder, latest_json_file)

print(f"📂 Loading the most recent raw data file: {latest_json_file}")
print(f"📁 Full path: {file_path}")

# load the JSON file into a Python dictionary
with open(file_path,'r',encoding='utf-8') as f:
    raw_data = json.load(f)

# Check the type and the top-level keys
print(f"\n✅ Data loaded successfully!")
print(f"   Type of 'raw_data': {type(raw_data)}")
print(f"   Top-level keys in the JSON: {raw_data.keys()}")



📂 Loading the most recent raw data file: raw_users_2026-08-05_12-29-46.json
📁 Full path: data/raw/raw_users_2026-08-05_12-29-46.json

✅ Data loaded successfully!
   Type of 'raw_data': <class 'dict'>
   Top-level keys in the JSON: dict_keys(['results', 'info'])


In [15]:
# The actual user data is inside the "results" key
users_list = raw_data['results']

# Let's check what we have
print(f"Total users fetched: {len(users_list)}")

# looking at the first user to understand the structure
first_user = users_list[0]
print(f"\nKeys inside a single user object: {first_user.keys()}")


Total users fetched: 5000

Keys inside a single user object: dict_keys(['gender', 'name', 'location', 'email', 'login', 'dob', 'registered', 'phone', 'cell', 'id', 'picture', 'nat'])


In [17]:
# flattening the nested dictionaries in to a flat table

df = pd.json_normalize(users_list,sep='_')

# Check the shape (rows, columns)
print(f"✅ DataFrame created successfully!")
print(f"   Shape (rows, columns): {df.shape}")
print(f"   Total columns: {df.shape[1]}")
print(f"\n📊 First 5 rows of the flattened DataFrame:")
df.head()

✅ DataFrame created successfully!
   Shape (rows, columns): (5000, 34)
   Total columns: 34

📊 First 5 rows of the flattened DataFrame:


,gender,email,phone,cell,nat,name_title,name_first,name_last,location_street_number,location_street_name,...,login_sha256,dob_date,dob_age,registered_date,registered_age,id_name,id_value,picture_large,picture_medium,picture_thumbnail
0,female,kubra.erbay@example.com,(875)-635-8712,(276)-197-7714,TR,Ms,Kübra,Erbay,2787,Istiklal Cd,...,e32100aff178434b60f541c5a1fadd1ee4cc22a814025d...,1964-11-14T11:43:52.159Z,61,2020-05-11T03:56:44.488Z,6,,NaN,https://randomuser.me/api/portraits/women/31.jpg,https://randomuser.me/api/portraits/med/women/...,https://randomuser.me/api/portraits/thumb/wome...
1,male,prhm.hmdy@example.com,069-74032840,0986-612-7065,IR,Mr,پرهام,احمدی,3758,حجاب,...,dbd968cef8e0bb259d0e3f76feb3eb63a7726544895d61...,1993-02-28T14:27:16.615Z,33,2008-12-19T01:31:56.308Z,17,,NaN,https://randomuser.me/api/portraits/men/98.jpg,https://randomuser.me/api/portraits/med/men/98...,https://randomuser.me/api/portraits/thumb/men/...
2,female,laurie.knight@example.com,O27 J25-3249,H23 D22-6792,CA,Mrs,Laurie,Knight,1065,Parliament St,...,3662e6d10b79c19eb71352d7c550783cbeb89ecd713a13...,1971-09-20T06:08:45.810Z,54,2021-12-29T01:06:24.484Z,4,SIN,221482581,https://randomuser.me/api/portraits/women/46.jpg,https://randomuser.me/api/portraits/med/women/...,https://randomuser.me/api/portraits/thumb/wome...
3,female,bettina.behr@example.com,0477-4230529,0170-6944336,DE,Mrs,Bettina,Behr,2372,Berliner Straße,...,28569a73102815dab1cdafbcf6b7d7be4e12ea990d32fc...,1957-01-02T10:37:17.478Z,69,2011-04-16T08:53:48.937Z,15,SVNR,55 020157 B 730,https://randomuser.me/api/portraits/women/74.jpg,https://randomuser.me/api/portraits/med/women/...,https://randomuser.me/api/portraits/thumb/wome...
4,female,aynz.rdyy@example.com,001-18412839,0913-180-2268,IR,Mrs,آیناز,رضایی,8259,ایران,...,429c463ca032b2d93bc66de4732a3ac19a26433321b9cc...,1984-01-04T03:16:11.272Z,42,2009-01-08T01:37:21.949Z,17,,NaN,https://randomuser.me/api/portraits/women/80.jpg,https://randomuser.me/api/portraits/med/women/...,https://randomuser.me/api/portraits/thumb/wome...


In [29]:
print("===== DATAFRAME INFO =====")
df.info()

print("\n===== STATISTICAL SUMMARY (Numerical) =====")
df.describe()

# Check for missing values (nulls) in each column.
print("\n===== MISSING VALUES COUNT PER COLUMN =====")
missing_values = df.isnull().sum()

# Only show columns that have at least 1 missing value to keep it clean
missing_values[missing_values > 0].sort_values(ascending = False)


===== DATAFRAME INFO =====
<class 'pandas.DataFrame'>
RangeIndex: 5000 entries, 0 to 4999
Data columns (total 34 columns):
 #   Column                          Non-Null Count  Dtype 
---  ------                          --------------  ----- 
 0   gender                          5000 non-null   str   
 1   email                           5000 non-null   str   
 2   phone                           5000 non-null   str   
 3   cell                            5000 non-null   str   
 4   nat                             5000 non-null   str   
 5   name_title                      5000 non-null   str   
 6   name_first                      5000 non-null   str   
 7   name_last                       5000 non-null   str   
 8   location_street_number          5000 non-null   int64 
 9   location_street_name            5000 non-null   str   
 10  location_city                   5000 non-null   str   
 11  location_state                  5000 non-null   str   
 12  location_country                

id_value    931
dtype: int64

In [31]:
df.describe()

,location_street_number,dob_age,registered_age
count,5000.000000,5000.000000,5000.000000
mean,4956.957800,53.213600,13.770000
std,2895.558755,16.398041,5.858939
min,1.000000,25.000000,4.000000
25%,2475.000000,39.000000,9.000000
50%,4920.000000,53.000000,14.000000
75%,7464.000000,67.000000,19.000000
max,9999.000000,81.000000,24.000000


## Initial Observations & Business Insights (Business Notes)

### 1. Data Volume & Structure
- **Total Records:** 5,000 user profiles.
- **Total Features:** 34 columns after flattening.
- **Good Coverage:** We have a wide range of attributes, including personal details (`name`, `gender`), contact info (`email`, `phone`), location (`street`, `city`, `country`, `timezone`), login credentials, and `dob`/`registered` dates.

### 2. Data Quality & Missing Values (Critical Business Risk)
- **Missing Data:** Only one column has significant missing values:
  - `id_value` (National ID / Passport / SSN) has **931 missing values** (18.6% of users).
  - 
- **Business Impact:** 
  - This is a **major KYC (Know Your Customer) risk**. 
  - Nearly 1 in 5 users signed up without providing a valid government ID. 
  - These users **cannot be fully onboarded** and will require manual verification, costing the company time and money.
  - **Recommendation:** Flag these 931 users for immediate "Profile Completion" email campaigns.

### 3. Data Types & Feature Extraction Opportunities
- **Date Columns:** `dob_date` and `registered_date` are stored as strings (`object`/`str`).
  - *Action:* Must convert to `datetime` format to extract month, year, and day-of-week trends.
- **Age Columns (Pre-calculated):** `dob_age` (age now) and `registered_age` (years since registration) are integers. These are already clean.
- **Email:** Currently a plain string. We can extract `email_domain` (e.g., `@gmail.com`, `@yahoo.com`, `@protonmail.com`) to detect disposable/fraudulent emails.
- **Location:** We have `location_country` and `location_city`. Great for geographic segmentation.

---

### 4. Statistical Insights (Numerical Columns)
Looking at `df.describe()`:

- **`dob_age` (Age):**
  - Range: **25 to 81 years old**.
  - Mean: ~53 years old. 
  - **Insight:** The platform skews toward an **older demographic** (middle-aged to senior). Marketing campaigns should target this age group, and UX should be designed with accessibility in mind.

- **`registered_age` (Years since registration):**
  - Range: **4 to 24 years**.
  - Mean: ~14 years.
  - **Insight:** Most users have been on the platform for **over a decade** (average 14 years). This indicates **high brand loyalty and retention**. 
  - *Correlation note:* An 81-year-old user who registered 24 years ago would have been 57 at sign-up. A 25-year-old user who registered 4 years ago would have been 21. This data aligns perfectly.

- **`location_street_number`:**
  - Range: 1 to 9999. Mean is ~4,957.
  - **Insight:** Looks like random generation. No major outliers to worry about here.

---

### 5. Hypothesis & Next Steps (EDA Focus)
1. **Hypothesis 1:** Users from certain countries have higher `id_value` null rates (e.g., developing nations may rely on alternative verification).
2. **Hypothesis 2:** Older users (`dob_age` > 60) might be responsible for the majority of missing `id_value` entries (digital literacy barrier).
3. **Hypothesis 3:** Users with disposable email domains (`@mailinator`, `@temp-mail`) are more likely to have null IDs (potential fraud).



---
*End of Initial Inspection*

---